### PR TITLE
Add TC CEPH-83573802

### PR DESCRIPTION
**Pull request by sayaleeraut**
Adding test case for CEPH-83573802 - Verify the cluster status when ongoing operation are stopped ( abrupt/crash) - Tier4

# Description
1. Bootstrap cluster  using cephadm    
2. Copy SSH keys to nodes and Add them to cluster with address and role labels attached to them    
3. Deploy all daemon services using cephadm   
4. Perform the same steps serially for all daemons and check ceph status:syntax: systemctl stop <ceph-daemon.service>
    Eg: systemctl stop ceph-osd@6.service
5. Check if the service is stopped
    # sytemctl status ceph-osd@6.service    
    The service should be stopped successfully
6. Verify the ceph status
    #ceph -s

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/cephadm/test_systemctl_ops_on_osd_service.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    pass
+
+LOG = Log(__name__) 
+
+def wait_for_cluster_health(node, status):
+    # Verify if the cluster is in healthy state
+    timeout, interval = 300, 20
+    for w in WaitUntil(timeout=timeout, interval=interval):
+        _status = Ceph(node).health()
+        if _status.split(' ', 1)[0] == status:
```
Comment (vamahaja) [inline tests/cephadm/test_systemctl_ops_on_osd_service.py:22]: use `_status.contains(_status)`

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/cephadm/test_osd_service_operations.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+class CephServiceStartError(Exception):
+    pass
+
+
+class CephClusterHealthError(Exception):
+    pass
+
+
+class ClientIsMissingError(Exception):
```
Comment (vamahaja) [inline tests/cephadm/test_osd_service_operations.py:18]: Replace this with - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L20

File: tests/cephadm/test_osd_service_operations.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+class CephServiceFailError(Exception):
+    pass
+
+
+class CephServiceStartError(Exception):
+    pass
+
+
+class CephClusterHealthError(Exception):
```
Comment (vamahaja) [inline tests/cephadm/test_osd_service_operations.py:14]: Replace this with - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L26

File: tests/cephadm/test_osd_service_operations.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+from cli.utilities.operations import wait_for_cluster_health
+from cli.utilities.utils import get_service_id, get_service_state, set_service_state
+from utility.log import Log
+
+
+class CephServiceFailError(Exception):
+    pass
+
+
+class CephServiceStartError(Exception):
```
Comment (vamahaja) [inline tests/cephadm/test_osd_service_operations.py:10]: Replace this with - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L32

File: tests/cephadm/test_osd_service_operations.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+from cli.utilities.operations import wait_for_cluster_health
+from cli.utilities.utils import get_service_id, get_service_state, set_service_state
+from utility.log import Log
+
+
+class CephServiceFailError(Exception):
```
Comment (vamahaja) [inline tests/cephadm/test_osd_service_operations.py:6]: Replace this with - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L32

File: cli/utilities/operations.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        node (ceph): Node to execute the cmd
+        status (str): Expected status, e.g. HEALTH_OK/HEALTH_WARN
+
+    Returns (bool): Based on expected v/s actual status
+    """
+    for w in WaitUntil(timeout=timeout, interval=interval):
+        _status = Ceph(node).health()
+        if status in _status:
+            log.info(f"Cluster status is in expected state {status}")
+            return True
```
Comment (vamahaja) [inline cli/utilities/operations.py:23]: Add extra line after for loop

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/cephadm/test_osd_service_operations.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+class CephClusterHealthError(Exception):
+    pass
+
+
+class ClientIsMissingError(Exception):
+    pass
+
+
+LOG = Log(__name__)
```
Comment (vamahaja) [inline tests/cephadm/test_osd_service_operations.py:22]: Follow same pattern, change variable name to `log`

File: tests/cephadm/test_osd_service_operations.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def run(ceph_cluster, **kw):
+    # Check if client node is present
+    client = ceph_cluster.get_nodes(role="client")
+    if not client:
+        raise ClientIsMissingError("Client node is missing, add a client node")
+    client = client[0]
```
Comment (vamahaja) [inline tests/cephadm/test_osd_service_operations.py:30]: More pythonic way -
```
client, osd_id = client[0], list()
```

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: suites/quincy/cephadm/tier-4-service-operations.yaml [hunk 9]
```diff
@@ -68,6 +68,20 @@ tests:
       module: test_cephadm.py
```
Comment (vamahaja) [inline suites/quincy/cephadm/tier-4-service-operations.yaml:68]: pacifci testsuite ?

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Comment by vamahaja** [comment]
Code looks to me, @MohitBis please review.

File: suites/quincy/cephadm/tier-4-service-operations.yaml (hunk 9 above)
Comment (sayaleeraut) [reply suites/quincy/cephadm/tier-4-service-operations.yaml:68]: Added for pacific as well

**Comment by vamahaja** [comment]
Test results -

- 5.3 - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1npob/
- 6.1 - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-jm0tj

**Pull request merged**